### PR TITLE
 Add as_nanos function to Duration

### PR DESCRIPTION
**Pull request by fintelia**
Duration has historically lacked a way to get the actual number of nanoseconds it contained as a normal Rust type because u64 was of insufficient range, and f64 of insufficient precision. The u128 type solves both issues, so I propose adding an `as_nanos` function to expose the capability.

### CODE REVIEW

File: src/libcore/time.rs [hunk 1]
```diff
@@ -263,6 +263,23 @@ impl Duration {
...
+    /// # Examples
+    ///
+    /// ```
+    /// # #![feature(duration_nanos)]
+    /// use std::time::Duration;
+    ///
+    /// let duration = Duration::new(5, 730023852);
+    /// assert_eq!(duration.as_nanos(), 5730023852);
+    /// ```
+    #[unstable(feature = "duration_nanos", issue = "50167")]
```
Comment (kennytm) [inline src/libcore/time.rs]: The tracking issue number should be a new issue, not this PR.

Comment (fintelia) [reply src/libcore/time.rs]: Done.

**Comment by pietroalbini** [comment]
Reassigning this to someone from the libs team.
r? @sfackler

**Comment by fintelia** [comment]
@sfackler I can make that change. Any thoughts on what the feature should be called then? "duration_as_nanos" isn't really a great name if it also adds as_millis() and as_micros(). Perhaps "duration_as_u128"?

**Comment by sfackler** [comment]
Sure, seems reasonable.

**Comment by pietroalbini** [comment]
Ping from triage @fintelia! It's been a while since we heard from you, will you have time to work on this again soon?

**Comment by fintelia** [comment]
Sorry about that. I'm probably going to be busy for about a week longer, so if someone else wants to take over making those changes, feel free

**Comment by pietroalbini** [comment]
@fintelia don't worry, we just want to make sure you still remember about the PR `;)`
Someone from triage will ping you next week.

File: src/libcore/time.rs [hunk 2]
```diff
@@ -263,6 +263,57 @@ impl Duration {
...
+    /// # #![feature(duration_as_u128)]
+    /// use std::time::Duration;
+    ///
+    /// let duration = Duration::new(5, 730023852);
+    /// assert_eq!(duration.as_millis(), 5730);
+    /// ```
+    #[unstable(feature = "duration_as_u128", issue = "50202")]
+    #[inline]
+    pub fn as_millis(&self) -> u128 {
+        self.secs as u128 * MILLIS_PER_SEC as u128 + self.nanos as u128 / NANOS_PER_MILLI as u128
```
Comment (clarcharr) [inline src/libcore/time.rs]: Nit: `self.nanos as u128 / NANOS_PER_MILLI as u128` should be replaced with `(self.nanos / NANOS_PER_MILLI) as u128`. The former requires doing 128-bit arithmetic on a 32-bit number, which is redundant.

File: src/libcore/time.rs [hunk 3]
```diff
@@ -263,6 +263,57 @@ impl Duration {
...
+    /// # #![feature(duration_as_u128)]
+    /// use std::time::Duration;
+    ///
+    /// let duration = Duration::new(5, 730023852);
+    /// assert_eq!(duration.as_micros(), 5730023);
+    /// ```
+    #[unstable(feature = "duration_as_u128", issue = "50202")]
+    #[inline]
+    pub fn as_micros(&self) -> u128 {
+        self.secs as u128 * MICROS_PER_SEC as u128 + self.nanos as u128 / NANOS_PER_MICRO as u128
```
Comment (clarcharr) [inline src/libcore/time.rs]: See comment from `as_nanos`

File: src/libcore/time.rs [hunk 4]
```diff
@@ -263,6 +263,57 @@ impl Duration {
...
+    /// # #![feature(duration_as_u128)]
+    /// use std::time::Duration;
+    ///
+    /// let duration = Duration::new(5, 730023852);
+    /// assert_eq!(duration.as_nanos(), 5730023852);
+    /// ```
+    #[unstable(feature = "duration_as_u128", issue = "50202")]
+    #[inline]
+    pub fn as_nanos(&self) -> u128 {
+        self.secs as u128 * NANOS_PER_SEC as u128 + self.nanos as u128
```
Comment (clarcharr) [inline src/libcore/time.rs]: See comment from `as_nanos`

**Comment by shepmaster** [comment]
Ping from triage, @sfackler !

**Comment by sfackler** [comment]
Oops - seems good to me!

@bors r+

**Pull request merged**